core/src/main/java/org/apache/iceberg/actions/RewriteDeleteStrategy.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.Map;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.Table;
+
+public interface RewriteDeleteStrategy {
+
+  /**
+   * Returns the table being modified by this rewrite strategy
+   */
+  Table table();
+
+  /**
+   * Select the deletes to rewrite.
+   *
+   * @return iterable of original delete file to be replaced.
+   */
+  Iterable<DeleteFile> selectDeletes();
```

Comment:
could you explain a bit about how these 2 interfaces will be used? For `RewriteStrategy` the method `Iterable<FileScanTask> selectFilesToRewrite(Iterable<FileScanTask> dataFiles)` was quite straightforward, but here the method has no arguments so I am not very sure how would it be used to actually select deletes.

I think it goes to the fundamental question of what are some potential rewrite delete strategies? For rewrite data files we know there are well-known operations like bin-packing. For deletes, what are some strategies we plan to implement?

Reply:
The argument of `selectFilesToRewrite(Iterable<FileScanTask> dataFiles)` comes from `planFileGroups` and the filter logic. I think we could build similar logic by storing the iterable of `FileScanTask` as a local variable in the concrete `RewrtiteDeleteStrategy` class, and grouping them in `rewriteDeletes()` API. As for filter logic, I think we can add it later and apply it in `selectDeletes()` API.

For strategies, I think it depends on different scenarios. In our cases, which mostly use HDFS as storage, we care more about the small files. they impact the planning, reading performance and also bring overhead to the name node. So the strategy we want badly is to merge deletes. This could be summarized to  1) merge all position deletes. 2) read all equality deletes under the partition and convert them into one position delete.  I think the first one is also bin-packing, the second one is convert-and-bin-packing.

Reply:
Thanks for the explanation. The case you described are also what I am interested in. I think for `RewriteStrategy` we first had a design doc so when there was the PR everything was in place. I am fine if we are taking incremental progress for delete strategies to gradually evolve the interface and add configuration options.

Reply:
Hi @chenjunjiedada ,  thanks for tracking the v2's compaction for long time, It's really challenging.   I think we reviewers were distracted by other internal or external things , which blocked the v2 compaction progress for a long time.  While the fact is: the current v2 compaction feature is the biggest blocker for v2 right now, I'd like to push this feature forward, so that we don't block v2 forever. (And thanks for your bandwidth @jackye1995,  you always come out to help when the things get stuck).

About this PR, since it's a API definition,  I'd like to take a look the whole delete rewrite design or preview PR if you have one because in that case I could evaluate whether the API abstraction is perfectly matching the real implementation.   Of course,  we could introduce a basic API, and then iterate the interface design when we get more implementation. It's looks better for me to reach a global agreement before we starting to provide a specific design,  that can help us reduce context sync on the core paths.

Reply:
+1 for a general design. And please let me know if there is anything that can be parallelized for implementation once we reached an agreement on this. I am currently mostly spending time trying to push the work on Trino side, but should have some bandwidth to take some implementation work for this if necessary. As openinx said, this is the biggest blocker for v2.

Reply:
I don't have a design doc right now, but most of the ideas come from discussions in https://github.com/apache/iceberg/pull/2216 and https://github.com/apache/iceberg/pull/2364. I'm working on rebasing #2216 to apply #2372 (API definition) and #2841 (delete row reader). That would be a full implementation and hopefully could give you the full picture of how these APIs are used.  I will post in the following days.

Reply:
great, let me read through those, thank you!